src/asymmetric_encryption.cpp

```diff
@@ -0,0 +1,242 @@
+/*
+ * #%L
+ * %%
+ * Copyright (C) 2018 BMW Car IT GmbH
+ * %%
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * #L%
+ */
+#include "mococrw/asymmetric_encryption.h"
+
+#include <fstream>
+
+#include "mococrw/error.h"
+#include "mococrw/hash.h"
+#include "mococrw/padding_mode.h"
+#include "mococrw/util.h"
+
+namespace mococrw {
+
+using namespace openssl;
+
+using SSL_RSA_OAEP_LABEL_Ptr = std::unique_ptr<uint8_t, SSLFree<uint8_t>>;
+
+using SSL_RSA_ENCRYPTION_DATA_Ptr = std::unique_ptr<unsigned char, SSLFree<unsigned char>>;
+
+AsymmetricEncryption::CryptoData::CryptoData(const std::string& data)
+    : _data{data.begin(), data.end()}
+{
+}
+
+AsymmetricEncryption::CryptoData::CryptoData(const std::vector<uint8_t>& data)
+    : _data{data}
+{
+}
+
+AsymmetricEncryption::CryptoData& AsymmetricEncryption::CryptoData::operator=(
+                                                                    const std::vector<uint8_t>& data)
+{
+    _data=data;
+    return *this;
+}
+
+AsymmetricEncryption::CryptoData& AsymmetricEncryption::CryptoData::operator=(
+                                                                    const std::string& data)
+{
+    _data=std::vector<uint8_t>(data.begin(), data.end());
+    return *this;
+}
+
+std::string AsymmetricEncryption::CryptoData::toString() const {
+    return std::string(_data.begin(), _data.end());
+}
+
+std::vector<uint8_t> AsymmetricEncryption::CryptoData::toByteArray() const {
+    return _data;
+}
+
+std::string AsymmetricEncryption::CryptoData::toHex() const {
+    return utility::toHex(_data);
+}
+
+std::ostream& operator<<(std::ostream& os, const AsymmetricEncryption::CryptoData& data)
+{
+    os << data.toString();
+    return os;
+}
+
+std::vector<uint8_t> AsymmetricEncryption::encrypt(AsymmetricPublicKey key,
+                                                   const RSAPadding &pad,
+                                                   const CryptoData& message)
+{
+    SSL_RSA_OAEP_LABEL_Ptr label_copy{nullptr};
+    SSL_RSA_ENCRYPTION_DATA_Ptr encryptedMessage{nullptr};
+    size_t encryptedMessageLen{0};
+    int maxSize{0};
+
+    try {
+        const auto keyCtx = _EVP_PKEY_CTX_new(key.internal());
+
+        const auto paddingMode = pad.getPadding();
+
+        if (!keyCtx.get()) {
+            throw MoCOCrWException("Encryption context is empty");
+        }
+
+        switch(paddingMode) {
```

Comment:
What if you move this logic into RSAPadding subclasses? Instead of asking for a padding mode and casting, just ask for a maximum supported message length.